### PR TITLE
Implement fn:format-integer#2,#3

### PR DESCRIPTION
**Pull request by alanpaxton**
### Description:

An implementation of fn:format-integer#2,#3
Makes use of ICU4J to format integers as words (https://unicode-org.github.io/icu/userguide/icu4j/)

This open source contribution to the eXist-db project was commissioned by the Office of the Historian, U.S. Department of State, https://history.state.gov.

### Reference:

https://www.w3.org/TR/xpath-functions-31/

### Type of tests:

Unit tests IntegerPictureTest tests covers the functionality from Java - developed in a TDD fashion.
Currently passes all XQTS tests in fn-format-integer (https://github.com/w3c/qt3tests/blob/master/fn/format-integer.xml)
except format-integer-052 - Kanji formatting is not implemented.

### CODE REVIEW

**Review by dizzzz — CHANGES_REQUESTED** [review]
build issue... missing file in commit? 
image

File: exist-core/src/main/java/org/exist/test/runner/XMLTestRunner.java [hunk 1]
```diff
@@ -78,18 +78,53 @@ public class XMLTestRunner extends AbstractTestRunner {
...
+     */
+    public static AbstractTestRunner getRunner(final Path path, final boolean parallel) throws InitializationError {
+
+        /*
+         * TODO (AP) pull this (factory) method out into AbstractTestRunner, merge with the getRunner() in XSuite.java
+         * so we end up with a number of clean sibling subclasses of AbstractTestRunner.
+         */
+        final Document doc;
+        final XMLTestInfo info;
+        final XQTSTestRunner.XQTSTestSuite xqtsTestSuite;
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/test/runner/XMLTestRunner.java:110]: please check scope

File: exist-core/src/main/java/org/exist/test/runner/XMLTestRunner.java [hunk 2]
```diff
@@ -133,19 +168,17 @@ private static XMLTestInfo extractTestInfo(final Path path, final Document doc)
...
-            throw new InitializationError("Could not find <testName> in XML <TestSet> document: " + path.toAbsolutePath().toString());
+            throw new InitializationError("Could not find <testName> in XML <TestSet> document: " + path.toAbsolutePath());
         }
 
         return new XMLTestInfo(testSetName, description, testNames);
     }
 
     private static @Nullable String getIdValue(final Node test) {
         String id = ((Element)test).getAttribute("id");
-        if (id != null) {
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/test/runner/XMLTestRunner.java:144]: are you sure there is no risk for a NPE ?

File: exist-core/src/main/java/org/exist/test/runner/XQTSTestRunner.java [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+import java.nio.file.Path;
+import java.util.*;
+
+/**
+ * A JUnit test runner which can run the XML formatted XQTS tests
+ * within eXist-DB.
+ *
+ * @author Alan Paxton
+ */
+public class XQTSTestRunner extends AbstractTestRunner {
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/test/runner/XQTSTestRunner.java:54]: is this a 'leakage' from a different PR? Mixup of concerns in PRs?

File: exist-core/src/main/java/org/exist/xquery/XQueryContext.java [hunk 4]
```diff
@@ -3317,6 +3317,14 @@ public void setSource(final Source source) {
         this.source = source;
     }
 
+    /**
+     * TODO (AP) this is a temporary implementation which will conflict and be sorted at merge
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/xquery/XQueryContext.java:3321]: ok

**Comment by dizzzz** [comment]
very valuable, need different PR first? which one?

File: exist-core/src/main/java/org/exist/test/runner/XMLTestRunner.java [hunk 5]
```diff
@@ -78,18 +78,53 @@ public class XMLTestRunner extends AbstractTestRunner {
...
+     * @throws InitializationError if the file doesn't contain a format of XML tests we know about
+     */
+    public static AbstractTestRunner getRunner(final Path path, final boolean parallel) throws InitializationError {
+
+        /*
+         * TODO (AP) pull this (factory) method out into AbstractTestRunner, merge with the getRunner() in XSuite.java
+         * so we end up with a number of clean sibling subclasses of AbstractTestRunner.
+         */
+        final Document doc;
+        final XMLTestInfo info;
```
[the remark this answers is not in the document]
Comment (alanpaxton) [reply exist-core/src/main/java/org/exist/test/runner/XMLTestRunner.java]: I've taken the test runner changes out; they will go into a separate PR.

File: exist-core/src/main/java/org/exist/test/runner/XMLTestRunner.java (hunk 1 above)
Comment (alanpaxton) [reply exist-core/src/main/java/org/exist/test/runner/XMLTestRunner.java]: I've taken the test runner changes out, they will go into a separate PR.

File: exist-core/src/main/java/org/exist/xquery/XQueryContext.java (hunk 4 above)
Comment (alanpaxton) [reply exist-core/src/main/java/org/exist/xquery/XQueryContext.java]: Temporary now removed.

**Comment by joewiz** [comment]
@dizzzz @adamretter Rather than consolidating the commits, I would propose that we simply "squash and merge" the PR when it's ready to be merged. The PR is focused and well-contained - adding support for a previously unsupported XQuery function - so this makes it a good candidate for squashing.

**Comment by adamretter** [comment]
@dizzzz @alanpaxton @joewiz I have done some rebasing on this branch to reduce it to about 1/2 the number of commits. I think some of the history in these commits is useful for anyone else that has to look at and understand how`fn:format-integer` was implemented in future.
In general I have tried to squash together similar commits, and I have also removed commits which added and then later removed files/changes.

Let me know if you are happy with this, or if you would still prefer to squash it into a single commit on merge?

**Comment by dizzzz** [comment]
OK for me, thnx

**Review by dizzzz — APPROVED** [review]
LGTM

**Pull request merged**